### PR TITLE
SmppNLSTSplitter for sending messages using National Language Shift Table

### PR DESCRIPTION
**Pull request by roadrunner**
to support sending messages with [National Language Shift Tables](https://en.wikipedia.org/wiki/GSM_03.38#National_language_shift_tables), there are two thing needs to be taken care of:
even if the short message is only one part, we need to set UDHI; we are using esm_class.
splitting messages for locking tables:
155 characters allowed when submitting single short message
149 characters allowed when submitting multipart short messages

i've added test cases for splitting

### CODE REVIEW

**Comment by roadrunner** [comment]
i wonder if you can apply this patch 2.18.x branch as well?

**Comment by oscerd** [comment]
Can you please Change the commit message? I think there is a typo :-)

**Comment by roadrunner** [comment]
Sorry, it was a typo :)

**Comment by oscerd** [comment]
Me or @davsclaus will review soon :-) thanks for your contribution!

**Comment by oscerd** [comment]
@davsclaus this looks good to me. What do you think?

File: components/camel-smpp/src/main/java/org/apache/camel/component/smpp/SmppNLSTSplitter.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
+package org.apache.camel.component.smpp;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Created by engin on 30/11/2016.
```
Comment (davsclaus) [inline components/camel-smpp/src/main/java/org/apache/camel/component/smpp/SmppNLSTSplitter.java]: Remove this kind of javadoc comments

File: components/camel-smpp/src/main/java/org/apache/camel/component/smpp/SmppNLSTSplitter.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
+package org.apache.camel.component.smpp;
```
Comment (davsclaus) [inline components/camel-smpp/src/main/java/org/apache/camel/component/smpp/SmppNLSTSplitter.java]: Add ASF license header

File: components/camel-smpp/src/main/java/org/apache/camel/component/smpp/SmppNLSTSplitter.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    protected static final int UDHIE_NLI_SINGLE_MSG_HEADER_LENGTH = 0x03; // header length for single message
+    protected static final int UDHIE_NLI_SINGLE_MSG_HEADER_REAL_LENGTH = UDHIE_NLI_SINGLE_MSG_HEADER_LENGTH + 1;
+
+    protected static final int UDHIE_NLI_MULTI_MSG_HEADER_LENGTH = 0x08; // header length for multi message
+    protected static final int UDHIE_NLI_MULTI_MSG_HEADER_REAL_LENGTH = UDHIE_NLI_MULTI_MSG_HEADER_LENGTH + 1;
+
+
+    protected static final int UDHIE_SAR_REF_NUM_LENGTH = 1;
+//        protected static final byte UDHIE_IDENTIFIER_SAR = 0x00;
```
Comment (davsclaus) [inline components/camel-smpp/src/main/java/org/apache/camel/component/smpp/SmppNLSTSplitter.java]: Remove commented out code, or explain why its commented out

**Comment by roadrunner** [comment]
i'm using this splitter in production without any issue. i can confirm that it's working for Turkish NLST is working. 
but i've to compile and install camel locally. i thought if i publish it others might contribute..

File: components/camel-smpp/src/main/java/org/apache/camel/component/smpp/SmppNLSTSplitter.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        int lengthOfData;
+        byte refNum = getReferenceNumber();
+        for (int i = 0; i < segmentNum; i++) {
+            logger.info("segment number = {}", i);
+            if (segmentNum - i == 1) {
+                lengthOfData = messageLength - i * segmentLength;
+            } else {
+                lengthOfData = segmentLength;
+            }
+            logger.info("Length of data = {}", lengthOfData);
```
Comment (davsclaus) [inline components/camel-smpp/src/main/java/org/apache/camel/component/smpp/SmppNLSTSplitter.java]: Log at DEBUG level

File: components/camel-smpp/src/main/java/org/apache/camel/component/smpp/SmppNLSTSplitter.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+            segments[i] = new byte[UDHIE_NLI_MULTI_MSG_HEADER_REAL_LENGTH + lengthOfData];
+            logger.info("segments[{}].length = {}", i, segments[i].length);
```
Comment (davsclaus) [inline components/camel-smpp/src/main/java/org/apache/camel/component/smpp/SmppNLSTSplitter.java]: Same here

File: components/camel-smpp/src/main/java/org/apache/camel/component/smpp/SmppSmCommand.java [hunk 6]
```diff
@@ -76,11 +76,19 @@ private SmppSplittingPolicy getSplittingPolicy(Message message) throws SmppExcep
...
-        Alphabet alphabet = determineAlphabet(message);
 
+        SmppSplitter splitter;
+        // use the splitter if provided via header
+        if (message.getHeaders().containsKey(SmppConstants.DATA_SPLITTER)){
+            splitter = message.getHeader(SmppConstants.DATA_SPLITTER, SmppSplitter.class);
+            if (null != splitter){
+                return splitter;
+            }
+            logger.warn("Invalid splitter given. Must be instance of SmppSplitter");
```
Comment (davsclaus) [inline components/camel-smpp/src/main/java/org/apache/camel/component/smpp/SmppSmCommand.java]: Why not throw an exception?

File: components/camel-smpp/src/test/java/org/apache/camel/component/smpp/SmppNLSTSplitterTest.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
+package org.apache.camel.component.smpp;
```
Comment (davsclaus) [inline components/camel-smpp/src/test/java/org/apache/camel/component/smpp/SmppNLSTSplitterTest.java]: ASF license header

File: components/camel-smpp/src/test/java/org/apache/camel/component/smpp/SmppNLSTSplitterTest.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Created by engin on 30/11/2016.
```
Comment (davsclaus) [inline components/camel-smpp/src/test/java/org/apache/camel/component/smpp/SmppNLSTSplitterTest.java]: Remove

**Comment by davsclaus** [comment]
And there is likely also some checkstyle issues to fix

**Comment by oscerd** [comment]
My comment was only about the feature. I didn't review the code :-)

**Comment by roadrunner** [comment]
okay, i'm dealing with the mess..

File: components/camel-smpp/src/main/java/org/apache/camel/component/smpp/SmppSmCommand.java (hunk 6 above)
Comment (roadrunner) [inline components/camel-smpp/src/main/java/org/apache/camel/component/smpp/SmppSmCommand.java]: +1 better to throw exception.  thanks for the tip

**Comment by davsclaus** [comment]
Looks better. @oscerd can you take a look as well and merge if okay.

**Comment by oscerd** [comment]
Sure :-)

**Comment by oscerd** [comment]
Thanks the PR has been merged. Can you close this?

**Comment by roadrunner** [comment]
Great, thank you guys..

**Comment by oscerd** [comment]
Thanks a lot for the contrib @roadrunner  :-)

**Comment by roadrunner** [comment]
i hope, i'll be doing more soon 👍

**Pull request closed without merging**